pkg/ccl/backupccl/backup_processor.go

```diff
@@ -312,23 +317,71 @@ func runBackupProcessor(
 	clusterSettings := flowCtx.Cfg.Settings
 
 	totalSpans := len(spec.Spans) + len(spec.IntroducedSpans)
-	todo := make(chan spanAndTime, totalSpans)
-	var spanIdx int
-	for _, s := range spec.IntroducedSpans {
-		todo <- spanAndTime{
-			spanIdx: spanIdx, span: s, firstKeyTS: hlc.Timestamp{}, start: hlc.Timestamp{},
-			end: spec.BackupStartTime,
+	requestSpans := make([]spanAndTime, 0, totalSpans)
+	rangeSizedSpans := preSplitExports.Get(&flowCtx.EvalCtx.Settings.SV)
+
+	for _, fullSpan := range spec.IntroducedSpans {
+		remainingSpan := fullSpan
+
+		if rangeSizedSpans {
+			rdi, err := flowCtx.Cfg.ExecutorConfig.(*sql.ExecutorConfig).RangeDescIteratorFactory.NewIterator(ctx, fullSpan)
+			if err != nil {
+				return err
+			}
+			for ; rdi.Valid(); rdi.Next() {
+				rangeDesc := rdi.CurRangeDescriptor()
+				subspan := remainingSpan.Intersect(roachpb.Span{Key: rangeDesc.StartKey.AsRawKey(), EndKey: rangeDesc.EndKey.AsRawKey()})
+
+				requestSpans = append(requestSpans, spanAndTime{
+					span: subspan, firstKeyTS: hlc.Timestamp{}, start: hlc.Timestamp{},
+					end: spec.BackupStartTime,
+				})
+				remainingSpan.Key = subspan.EndKey
+			}
+
+			if !remainingSpan.Overlaps(fullSpan) {
+				continue
+			}
 		}
-		spanIdx++
+
+		requestSpans = append(requestSpans, spanAndTime{
+			span: remainingSpan, firstKeyTS: hlc.Timestamp{}, start: hlc.Timestamp{},
+			end: spec.BackupStartTime,
+		})
 	}
-	for _, s := range spec.Spans {
-		todo <- spanAndTime{
-			spanIdx: spanIdx, span: s, firstKeyTS: hlc.Timestamp{}, start: spec.BackupStartTime,
-			end: spec.BackupEndTime,
+
+	for _, fullSpan := range spec.Spans {
+		remainingSpan := fullSpan
+
+		if rangeSizedSpans {
+			rdi, err := flowCtx.Cfg.ExecutorConfig.(*sql.ExecutorConfig).RangeDescIteratorFactory.NewIterator(ctx, fullSpan)
+			if err != nil {
+				return err
+			}
+			for ; rdi.Valid(); rdi.Next() {
+				rangeDesc := rdi.CurRangeDescriptor()
+				subspan := remainingSpan.Intersect(roachpb.Span{Key: rangeDesc.StartKey.AsRawKey(), EndKey: rangeDesc.EndKey.AsRawKey()})
+
+				requestSpans = append(requestSpans, spanAndTime{
+					span: subspan, firstKeyTS: hlc.Timestamp{}, start: hlc.Timestamp{},
+					end: spec.BackupStartTime,
+				})
+				remainingSpan.Key = subspan.EndKey
+			}
+
+			if !remainingSpan.Overlaps(fullSpan) {
+				continue
+			}
```

Comment:
NB: If I'm reading correctly we expect this to happen on the laster iteration because we will have set our remainingSpan.Key to the EndKey of the last range. So either we have a single key span that won't overlap fullSpan or we have an inverted range.

Reply:
Are you suggesting this should be if `!remainingSpan.Valid() { break }` instead?

Reply:
I was mostly just confirming for myself what this code was doing.  Is Valid() the correct condition here? If Key == EndKey isn't Valid() still true but we likely don't want to add the single-key span?

Reply:
Key == EndKey isn't a single key span is it? EndKey is exclusive so that's an empty span isn't it?

Reply:
Ah, I misread the Valid() function, equal keeps indeed returns false.